# SIMPLISTIC INTRODUCTION TO ALGORITHMIC COMPLEXITY
<hr style="height:2px;color:blue"/>
The most important thing to think about when designing and implementing a program is 

that it should produce <b>results</b> that can <b>be relied upon</b>.

Sometimes <b>performance</b> is an important aspect of <b>correctness</b>.

  * This is most obvious for programs that need to <b>run in real time</b>
  

  * Performance can also affect <b>the utility of many non-real-time programs</b>

Programmers often <b>increase</b> the <b>conceptual complexity</b> of a program in an effort to <b>reduce</b> its <b>computational complexity</b>.

To do this in a <b>sensible</b> way, 

we need to understand how to go about <b>estimating the computational complexity</b> of a program.

## 1 Thinking About Computational Complexity

How should one go about answering the question

* **How long will the following function take to run?**


In [ ]:
def f(i):
    """Assumes i is an int and i >= 0"""
    answer = 1
    while i >= 1:
        answer *= i
        print(answer)
        i -= 1
    return answer

We could run the program on some input and <b>time</b> it. 

The <b>result would depend upon</b>

* the speed of the <b>computer</b> on which it is run,

* the efficiency of the <b>Python implementation</b> on that machine, and

* the value of <b>the input<b>.


**We get around the first two issues**

*  </b> by using <b>a more abstract measure of time</b>. 

We measure <b>time</b> in terms of 
  
*  <b>the number of basic steps</b> 
  
executed by the program.

For simplicity, we will use <b>a random access machine</b> as our model of
computation. 

https://en.wikipedia.org/wiki/Random-access_machine

In a random access machine：

* <b>steps are executed sequentially</b>, <b>one at a time</b>. 

* A <b>step</b> is an operation that takes <b>a fixed amount of time</b>.

Now that we have <b>a more abstract way</b> to think about the meaning of time,

we turn to the question of **dependence on the value of the `input`**. 

We deal with that by expressing time complexity as relating it to 

* <b>the sizes of the inputs</b>. 

This allows us to compare the efficiency of two algorithms by talking about

* <b>how the running time of each grows with respect to the sizes of the inputs</b>

**The `linear search algorithm` implemented by**：


In [ ]:
def linearSearch(L, x):
    for e in L:
        if e == x:
            return True
    return False

Suppose that L is a million elements long 

consider the call

```python
  linearSearch(L, 3).
```
If the first element in L is 3, linearSearch will return True almost immediately.

if 3 is not in L, linearSearch will have to examine all one million elements before returning False.

In general, there are <b>three broad cases</b> to think about：

* the <b>best-case</b> running time is <b>the minimum running time</b> over all the possible inputs of a given size.

   For linearSearch, the best-case running time is <b>independent of the size of L</b>.
   
*  the <b>worst-case</b> running time is <b>the maximum running time</b> over all the possible inputs of a given size. 

   For linearSearch, the worstcase running time is <b>linear in the size of the list</b>.

*  the <b>average-case</b> (also called expected-case) running time is the average running time over all possible inputs of a given size. 


People usually focus <b>on the worst case</b> :an <b>upper bound</b> on the running time. 

This is <b>critical</b> in situations where there is <b>a time constraint</b> on how long a computation can take. 


Let’s look at the <b>worst-case running time</b> of an iterative implementation of the factorial function


In [ ]:
def fact(n):
    """Assumes n is a natural number
    Returns n!"""
    answer = 1
    while n > 1:
        answer *= n
        n -= 1
    return answer

The number of <b>steps</b> required to run this program is something like：<b> 2+5*n</b>

* **2**：
      answer = 1    # for the initial assignment statement
      return answer #for the return

* **5*n**：
   
   * **1** step for the test in the while： while n > 1:
      
   * **2** steps for the first assignment statement in the while loop： answer *= n
      
   * **2** steps for the second assignment statement in the loop： n -= 1
      
<b>Multiplicative</b> factors can be <b>important</b>.

On the other hand, when one is comparing two <b>different algorithms</b>, 

it is often the case that even <b>multiplicative constants</b> are <b>irrelevant</b>.

* ** squareRoot**

  * ** Exhaustive**

  * ** BiSection**

In [2]:
def squareRootExhaustive(x, epsilon):
    """Assumes x and epsilon are positive floats & epsilon < 1
    Returns a y such that y*y is within epsilon of x"""
    step = epsilon**2
    ans = 0.0
    
    while abs(ans**2 - x) >= epsilon and ans*ans <= x:
        ans += step
  
    if ans*ans > x:
        raise ValueError
    return ans

In [3]:
def squareRootBi(x, epsilon):
    """Assumes x and epsilon are positive floats & epsilon < 1
    Returns a y such that y*y is within epsilon of x"""
    low = 0.0
    high = max(1.0, x)
    ans = (high + low)/2.0
    iterations=0
   
    while abs(ans**2 - x) >= epsilon:
        
        if ans**2 < x:
            low = ans
        else:
            high = ans
            
        ans = (high + low)/2.0
   
        iterations +=1
   
    return ans,iterations

In [4]:
ans,iterations=squareRootBi(100, 0.0001)
print('ans=',ans)
print('iterations=',iterations)


ans= 10.000002384185791
iterations= 22


**squareRootExhaustive(100, 0.0001)** 

   requires roughly one billion iterations of the loop. 

**squareRootBi(100, 0.0001)** 

   takes roughly twenty iterations of a slightly more complex while loop.

When the difference in <b>the number of iterations is this large</b>, 

it doesn’t really matter how many instructions are in the loop. 

I.e., the multiplicative constants are irrelevant.

## 2 Asymptotic Notation


In [6]:
def f(x):
    """Assume x is an int > 0"""
    ans = 0
   
   #1 Loop that takes constant time:1000
    for i in range(1000):
        ans += 1
    print('Number of additions so far', ans)
   
    #2 Loop that takes time:  x
    for i in range(x):
        ans += 1
    print('Number of additions so far', ans)
   
    #3 Nested loops take time x**2+ x**2
    for i in range(x):
        for j in range(x):
            
            ans += 1
            
            ans += 1
    
    print('Number of additions so far', ans)
   
    return ans

If one assumes that each line of code takes one unit of time to execute, the running time of this function can be described as 

<b>1000 + x + 2*x^2<b>.

In [7]:
f(10)

Number of additions so far 1000
Number of additions so far 1010
Number of additions so far 1210


1210

If x is 10, 

1000 of the 1210 steps are accounted for by the first loop.

if x is 1000, 

each of the first two loops accounts for only 0.05% of the steps.

if x is 1,000,000, 

the first loop takes about 0.00000005% of the total time and 

the second loop about 0.00005%. 

A full 2,000,000,000,000 steps are in the body of the inner for loop

This kind of analysis leads us to use the following rules of 

<b>thumb in describing the asymptotic complexity</b> of an algorithm:

* If the running time is <b>the sum of multiple terms</b>, 

    <b>keep the one with the largest growth rate</b>, and drop the others.


* If the remaining term is <b>a product</b>,

   <b>drop any constants</b>.

We use something called <b>asymptotic notation</b> to provide <b>a formal way</b> to talk about 

<b>the relationship between the running time of an algorithm and the size of its inputs</b>.

<b>asymptotic notation</b> as <b>a proxy for “very large”</b>, 

describes the complexity of an algorithm as the size of its inputs approaches infinity.

<b>The most commonly used asymptotic notation is called “Big O” notation</b>.

<b>$Big O$</b> notation is used to give <b>an upper bound</b> :                                                                                                                                                                                  called <b>the order of growth</b>) of a function.

For example, the formula 

$$f(x) ∈ O(x^2)$$ 
 
means that the function $f$ grows no faster than the quadratic polynomial x^2, in an asymptotic sense.

The difference between a function being <b>“in O(x^2)” and “being O(x^2)”</b> is subtle but important. 

**in O(x^2)**: Saying that $f(x) ∈ O (x^2)$ does not preclude the **worst-case** running time of f from being considerably **less** that O(x^2).

**being O(x^2)**: When we say that $f(x)$ is $O(x^2)￥, 

  we are implying that x^2 is both an **upper and a lower bound** on the asymptotic **worst-case** running time. This is called a **tight bound**.


## 3 Some Important Complexity Classes

Some of the most common instances of <b>Big O</b> are listed below. In each case, <b>n</b> is a measure of the size of the inputs to the function.

* **$O(1$)** denotes constant running time.



* **$O(logn)$** denotes logarithmic running time.


* **$O(n)$** denotes linear running time.

* **$O(nlogn)$** denotes log-linear running time.


* **$O(n^k)$** denotes polynomial running time. Notice that k is a constant.


* **$O(c^n)$** denotes exponential running time. Here a constant is being raised to a power based on the size of the input.


### 3.1 Constant Complexity

This indicates that the asymptotic complexity is <b>independent of the inputs</b>

### 3.2 Logarithmic Complexity

Such functions have a complexity that grows as <b>the log of at least one of the inputs</b>.


In [3]:
def intToStr(i):
    """Assumes i is a nonnegative int
      Returns a decimal string representation of i"""
    digits = '0123456789'
    if i == 0:
        return '0'
    result = ''
   
    while i > 0:
        result = digits[i%10] + result
        i = i//10  
  
    return result

That boils down to the number of times one can divide i by 10. So, the complexity of intToStr is <b>O(log(i))</b>.

In [4]:
def addDigits(n):
    """Assumes n is a nonnegative int
      Returns the sum of the digits in n"""
    stringRep = intToStr(n)
    val = 0
    for c in stringRep:
        val += int(c)
    return val,len(stringRep)

In [5]:
val,lenstringRep=addDigits(1376)

In [6]:
print(val)
print(lenstringRep)

17
4


The complexity of converting n to a string is O(log(n)) and intToStr returns a string of length O(log(n)).

the program will run in time proportional to O(log(n)) + O(log(n)), which makes it O(log(n)).

### 3.3 Linear Complexity

Many algorithms that deal with <b>lists or other kinds of sequences</b> are <b>linear</b> because they touch each element of the sequence <b>a constant</b> (greater than 0) number of times.



In [7]:
def addDigits(s):
    """Assumes s is a str each character of which is a
         decimal digit.
      Returns an int that is the sum of the digits in s"""
    val = 0
    for c in s:
        val += int(c)
    
    return val


This function is linear in the length of $s$.

Of course, a program does not need to have a <b>loop</b> to have <b>linear</b> complexity.

In [ ]:
def factorial(x):
    """Assumes that x is a positive int
      Returns x!"""
    if x == 1:
        return 1
    else:
        return x*factorial(x-1)

At the maximum depth of recursion, this code will have allocated x stack frames, so the <b>space complexity</b> is O(x).

The impact of space complexity is <b>harder</b> to appreciate than the impact of time complexity.

### 3.4 Log-Linear Complexity

$O(n log(n))$: It involves the <b>product of two terms</b>, each of which depends upon the size of the inputs.

The most commonly used log-linear algorithm is probably <b>merge sort</b>, which is <b>$O(n log(n))$</b>, where n is the length of the list being sorted.


### 3.5 Polynomial Complexity

The most commonly used polynomial algorithms are <b>quadratic</b>, i.e., 

* their complexity grows as the square of the size of their input:$O(n^2）$.

In [8]:
def isSubset(L1, L2):
    """Assumes L1 and L2 are lists.
      Returns True if each element in L1 is also in L2
      and False otherwise."""
    for e1 in L1:
        matched = False
        for e2 in L2:
            if e1 == e2:
                matched = True
            break
        if not matched:
            return False
   
    return True

the complexity of isSubset is $O(len(L1)*len(L2))$.

In [9]:
def intersect(L1, L2):
    """Assumes: L1 and L2 are lists
      Returns a list that is the intersection of L1 and L2"""
    #Build a list containing common elements
    tmp = []
    for e1 in L1:     # O(len(L1))
        for e2 in L2:  # O(len(L2))
            if e1 == e2:
                tmp.append(e1)
    #Build a list without duplicates
    result = []
    for e in tmp:          # O(len(tmp))
        if e not in result: # O(len(result))
            result.append(e)
   
    return result

Build a list containing common elements： The running time for the part building the list that might contain duplicates is clearly O(len(L1)*len(L2)).

Build a list without duplicates： $O(len(tmp)*len(result))$.

the complexity of intersect is $O(len(L1)*len(L2))$.





### 3.6 Exponential Complexity

As we will see later in this book, many important problems are inherently exponential, i.e., solving them completely can require time that is exponential in the size of the input.

This is unfortunate, since it rarely pays to write a program that has a reasonably high probability of taking exponential time to run.

In [10]:
def getBinaryRep(n, numDigits):
    """Assumes n and numDigits are non-negative ints
      Returns a numDigits str that is a binary
      representation of n"""
    result = ''
    while n > 0:
        result = str(n%2) + result
        n = n//2
    
    if len(result) > numDigits:
        raise ValueError('not enough digits')
   
    for i in range(numDigits - len(result)):
        result = '0' + result
    return result


In [11]:
print(getBinaryRep(11,4))

1011


In [ ]:
print(getBinaryRep(11,3))

In [ ]:
print(getBinaryRep(11,5))

In [12]:
def genPowerset(L):
    """Assumes L is a list
      Returns a list of lists that contains all possible
      combinations of the elements of L.  E.g., if
      L is [1, 2] it will return a list with elements
      [], [1], [2], and [1,2]."""
    powerset = []
    for i in range(0, 2**len(L)):
        binStr = getBinaryRep(i, len(L))
        subset = []
        for j in range(len(L)):
            if binStr[j] == '1':
                subset.append(L[j])
        # demo L=['a', 'b']
        print(binStr)
        print(subset)
    
        powerset.append(subset)
    
    return powerset

The function ```genPowerset(L)``` returns <b>a list of lists</b> that contains all possible combinations of the elements of L. 

For example, if L is ['a', 'b'], the powerset of L will be a list containing the lists [], ['b'], ['a'], and ['a','b'].
 
Consider a list of n elements. We can represent any combination of elements by a string of n 0’s and 1’s, where a 1 represents the presence of an element and a 0 its absence.

Therefore generating all sublists of a list L of length n can be done as follows:

1. Generate all n-bit binary numbers. These are the numbers from <b>0 to 2^n</b>.
```python
 for i in range(0, 2**len(L)):
      binStr = getBinaryRep(i, len(L))
```

2. For each of these 2n +1 binary numbers, b, generate a list by selecting    those elements of L that have <b>an index </b>corresponding to a 1 in b. 

In [ ]:
L=['a', 'b']
genPowerset(L)

For example, if L is ['a', 'b'] and b is 01, generate the list [‘b’].

Try running ```genPowerset``` on a list containing the four letters of the alphabet. It will finish quite quickly and produce a list with 16(2^4) elements.

In [ ]:
L=['a', 'b','c', 'd'] 
powerset=genPowerset(L)
print(len(powerset))

In [ ]:
L=['aa', 'bc',13]
powerset=genPowerset(L)
print(powerset)
print(len(powerset))


<b>Step 1</b> of the algorithm generates $O(2^len(L))$ binary numbers, so the algorithm is exponential in $len(L)$.

### 3.7 Comparisons of Complexity Classes

The following plots are intended to convey an impression of the implications of an algorithm being in one or another of these complexity classes.

The plot compares the growth of a <b>constant-time</b> algorithm to that of a <b>logarithmic</b> algorithm.
<img src="./img/ds/9.3.7.1.PNG"/> 
The moral is that logarithmic algorithms are<b>almost as good as</b> constant-time ones.


The plot illustrates the dramatic difference between <b>logarithmic algorithms </b> and  <b>linear algorithms </b>.
<img src="./img/ds/9.3.7.2.PNG"/> 
The difference between logarithmic-time and linear-time algorithms is apparent even on small inputs

Most of the time a linear algorithm is acceptably efficient.

The plot shows that there is a significant difference between Linear<b>O(n)</b> and <b>log-linear algorithm O(n log(n))</b>.
<img src="./img/ds/9.3.7.3.PNG"/> 
Given how slowly <b>log(n) grows</b>, this may seem a bit surprising, but keep in mind that it is <b>a multiplicative factor</b>. 
Also keep in mind that in most practical situations, O(<b>n</b> log(n)) is <b>fast enough </b>to be useful.

On the other hand, as the plot suggests, there are many situations in which <b>a quadratic</b> rate of growth is <b>prohibitive</b>. 

The quadratic curve is rising so quickly that it is hard to see that the <b>log-linear</b> curve is even on the plot.
<img src="./img/ds/9.3.7.4.PNG"/> 

The final two plots are about exponential complexity. 

The plot compares the growth of a <b>quadratic</b> of growth algorithm to that of a <b>exponential</b> algorithm 
<img src="./img/ds/9.3.7.5.PNG"/> 
In the plot on the left, the numbers to the left of the y-axis run from 0.0 to 1.2.

However, the notation $*1e301$ on the top left means that each tick on the y-axis  should be multiplied by $10^{301}$. 

So, the plotted y-values range from 0 to roughly  $1.1*10301$. It looks, however, almost as if there are no curves in the plot on the
left.

The plot on the right addresses this issue by using a <b>logarithmic scale on the y-axis</b>. 
<img src="./img/ds/9.3.7.6.PNG"/> 
One can readily see that exponential algorithms are impractical for all but the smallest of inputs.